refactor(core): DI Cleanup

mhevery opened the pull request:
refactor(core): Cleanup circular dependency between ViewEngine and Ivy.

`*Ref` is declared in ViewEngine but it sub-classed in Ivy. This creates a circular dependency between ViewEngine `*Ref` which needs to declare `__NG_ELEMENT_ID__` and ivy factory which needs to create it. The workaround used to be to pass the `*Ref` through stack but that created a very convoluted code. This refactoring simply bundles the two files together and removes the stack workaround making the code simpler to follow.

refactor(core): Cleanup non-standard `Injector` handling.
    
Due to historical reasons `Injector.__NG_ELEMENT_ID__` was set to `-1`. This changes it to be consistent with other `*Ref.__NG_ELEMENT_ID__` constructs.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [X] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mhevery committed:
refactor(core): Cleanup circular dependency between ViewEngine and Ivy `ElementRef`.

`ElementRef` is declared in ViewEngine but it sub-classed in Ivy. This creates a circular
dependency between ViewEngine `ElementRef` which needs to declare `__NG_ELEMENT_ID__` and
ivy factory which needs to create it. The workaround used to be to pass the `ElementRef`
through stack but that created a very convoluted code. This refactoring simply bundles the
two files together and removes the stack workaround making the code simpler to follow.

mhevery committed:
refactor(core): Cleanup circular dependency between ViewEngine and Ivy `TemplateRef`.

`TemplateRef` is declared in ViewEngine but it sub-classed in Ivy. This creates a circular
dependency between ViewEngine `TemplateRef` which needs to declare `__NG_ELEMENT_ID__` and
ivy factory which needs to create it. The workaround used to be to pass the `TemplateRef`
through stack but that created a very convoluted code. This refactoring simply bundles the
two files together and removes the stack workaround making the code simpler to follow.

mhevery committed:
refactor(core): Cleanup circular dependency between ViewEngine and Ivy `ViewContainerRef`.

`ViewContainerRef` is declared in ViewEngine but it sub-classed in Ivy. This creates a circular
dependency between ViewEngine `ViewContainerRef` which needs to declare `__NG_ELEMENT_ID__` and
ivy factory which needs to create it. The workaround used to be to pass the `ViewContainerRef`
through stack but that created a very convoluted code. This refactoring simply bundles the
two files together and removes the stack workaround making the code simpler to follow.

mhevery committed:
refactor(core): Cleanup circular dependency between ViewEngine and Ivy `ChangeDetectorRef`.

`ChangeDetectorRef` is declared in ViewEngine but it sub-classed in Ivy. This creates a circular
dependency between ViewEngine `ChangeDetectorRef` which needs to declare `__NG_ELEMENT_ID__` and
ivy factory which needs to create it. The workaround used to be to pass the `ChangeDetectorRef`
through stack but that created a very convoluted code. This refactoring simply bundles the
two files together and removes the stack workaround making the code simpler to follow.

mhevery committed:
refactor(core): Cleanup circular dependency between ViewEngine and Ivy `Renderer2`.

`Renderer2` is declared in ViewEngine but it sub-classed in Ivy. This creates a circular
dependency between ViewEngine `Renderer2` which needs to declare `__NG_ELEMENT_ID__` and
ivy factory which needs to create it. The workaround used to be to pass the `Renderer2`
through stack but that created a very convoluted code. This refactoring simply bundles the
two files together and removes the stack workaround making the code simpler to follow.

mhevery committed:
refactor(core): Cleanup non-standard `Injector` handling.

Due to historical reasons `Injector.__NG_ELEMENT_ID__` was set to `-1`.
This changes it to be consistent with other `*Ref.__NG_ELEMENT_ID__`
constructs.

petebacondarwin reviewed (approved):
Not an area of expertise for me, but a high level review looks good to me.
I am somewhat concerned by the very large increase in circular dependencies... aren't we trying to minimize these?

mhevery commented:
> Not an area of expertise for me, but a high level review looks good to me.
> I am somewhat concerned by the very large increase in circular dependencies... aren't we trying to minimize these?

Yes, I am going to look into it see if there could be anything done.

mhevery committed:
refactor(core): extract `DoBootstrap` to separate file.

Extract `DoBootstrap` interface to a separate file to break circular dependency.

mhevery committed:
refactor(core): Remove circular dependency between `LContainer` and `ViewRef`.

`LContainer` stores `ViewRef`s this is not quite right as it creates
circular dependency between the two types. Also `LContainer` should not
be aware of `ViewRef` which iv ViewEngine specific construct.

mhevery committed:
refactor(core): Remove circular dependency on `render3` and `ng_module`

Extracted `NgModeDef` into a separate file to break the circular dependency.

AndrewKushnir commented on packages/core/src/linker/view_container_ref.ts:
```diff
@@ -148,11 +168,266 @@ export abstract class ViewContainerRef {
+    }
+  }
+
+  detach(index?: number): ViewRef|null {
+    this.allocateContainerIfNeeded();
+    const adjustedIdx = this._adjustIndex(index, -1);
+    const view = detachView(this._lContainer, adjustedIdx);
+
+    const wasDetached = view && removeFromArray(this._lContainer[VIEW_REFS]!, adjustedIdx) != null;
+    return wasDetached ? new R3ViewRef(view!) : null;
+  }
```
I'd like to propose a small refactoring here:
```suggestion
  detach(index?: number): ViewRef|null {
    const lContainer = this._lContainer;
    const adjustedIdx = this._adjustIndex(index, -1);
    const view = detachView(lContainer, adjustedIdx);

    const viewRefs = getOrCreateViewRefs(lContainer);  // <-- previously this.allocateContainerIfNeeded()
    const wasDetached = view && removeFromArray(viewRefs, adjustedIdx) != null;
    return wasDetached ? new R3ViewRef(view!) : null;
  }
```

and the `getOrCreateViewRefs` (replacement to `allocateContainerIfNeeded`) might looks like:

```
  function getOrCreateViewRefs(lContainer: LContainer): ViewRef[] {
    if (lContainer[VIEW_REFS] === null) {
      lContainer[VIEW_REFS] = [];
    }
    return lContainer[VIEW_REFS];
  }
```

This is outside of the main scope of this PR, so we can do that later too.

AndrewKushnir commented on packages/core/src/linker/view_container_ref.ts:
```diff
@@ -148,11 +168,266 @@ export abstract class ViewContainerRef {
+ */
+export function createContainerRef(
+    hostTNode: TElementNode|TContainerNode|TElementContainerNode,
+    hostView: LView): ViewContainerRef {
```
```suggestion
    hostLView: LView): ViewContainerRef {
```

mhevery replied:
fixed

AndrewKushnir commented on packages/core/src/linker/view_container_ref.ts:
```diff
@@ -148,11 +168,266 @@ export abstract class ViewContainerRef {
+      // this specific case we use low-level DOM manipulation to insert container's marker (comment)
+      // node.
+      if (isRootView(hostView)) {
+        const renderer = hostView[RENDERER];
```
It looks like we can move this line higher above line 403 where we do `hostView[RENDERER].createComment`.

mhevery replied:
fixed

AndrewKushnir commented on packages/core/src/render/api.ts:
```diff
@@ -286,6 +290,26 @@ export abstract class Renderer2 {
+const SWITCH_RENDERER2_FACTORY: typeof injectRenderer2 = SWITCH_RENDERER2_FACTORY__PRE_R3__;
+
+/** Returns a Renderer2 (or throws when application was bootstrapped with Renderer3) */
+function getOrCreateRenderer2(view: LView): Renderer2 {
```
```suggestion
function getOrCreateRenderer2(lView: LView): Renderer2 {
```

mhevery replied:
fixed

AndrewKushnir commented on packages/core/src/render/api.ts:
```diff
@@ -286,6 +290,26 @@ export abstract class Renderer2 {
+  const renderer = view[RENDERER];
+  if (isProceduralRenderer(renderer)) {
+    return renderer as Renderer2;
+  } else {
+    throw new Error('Cannot inject Renderer2 when the application uses Renderer3!');
+  }
```
We can probably use `ngDevMode` here:
```suggestion
  } else if (ngDevMode) {
    throw new Error('Cannot inject Renderer2 when the application uses Renderer3!');
  }
```

mhevery replied:
fixed

AndrewKushnir commented on packages/core/src/render3/view_engine_compatibility_prebound.ts:
```diff
@@ -26,7 +22,7 @@ import {createTemplateRef, injectChangeDetectorRef} from './view_engine_compatib
  * @codeGenApi
  */
 export function ɵɵtemplateRefExtractor(tNode: TNode, currentView: LView) {
```
```suggestion
export function ɵɵtemplateRefExtractor(tNode: TNode, currentLView: LView) {
```

mhevery replied:
fixed

AndrewKushnir reviewed (commented):
That looks great, thanks @mhevery 👍 

FYI, I've left some comments and most of them are for pre-existing code that got moved (so it showed up in the diff). Feel free to ignore the ones that are not directly related to this PR, we can do more cleanup later in a followup PR(s).

Thank you.

mhevery committed:
refactor(core): Remove circular dependency on `render3` JIT and ViewEngine

JIT needs to identify which type is `ChangeDetectorRef`. It was doing so
by importing `ChangeDetectorRef` and than comparing the types. This creates
circular dependency as well as prevents tree shaking. The new solution is
to brand the class with `__ChangeDetectorRef__` so that it can be identified
without creating circular dependency.

mhevery committed:
refactor(core): Remove circular dependency on `ApplicationRef`

`ViewRef` and `ApplicationRef` had a circular reference. This change
introduces `ViewRefTracker` which is a subset of `ApplicationRef` for
this purpose.

mhevery commented on packages/core/src/di/null_injector.ts:
```diff
@@ -0,0 +1,25 @@
+export class NullInjector implements Injector {
+  get(token: any, notFoundValue: any = THROW_IF_NOT_FOUND): any {
+    if (notFoundValue === THROW_IF_NOT_FOUND) {
+      // Intentionally left behind: With dev tools open the debugger will stop here. There is no
+      // reason why correctly written application should cause this exception.
```
removed

mhevery commented on packages/core/src/di/null_injector.ts:
```diff
@@ -0,0 +1,25 @@
+    if (notFoundValue === THROW_IF_NOT_FOUND) {
+      // Intentionally left behind: With dev tools open the debugger will stop here. There is no
+      // reason why correctly written application should cause this exception.
+      const error = new Error(`NullInjectorError: No provider for ${stringify(token)}!`);
```
This would require changing the tests because the text is not same. Would prefer to do it in separate PR.

mhevery commented on packages/core/src/linker/element_ref.ts:
```diff
@@ -6,9 +6,38 @@
+/**
+ * Creates an ElementRef given a node.
+ *
+ * @param ElementRefToken The ElementRef type
```
fixed

mhevery commented on packages/core/src/linker/element_ref.ts:
```diff
@@ -6,9 +6,38 @@
+ *
+ * @param ElementRefToken The ElementRef type
+ * @param tNode The node for which you'd like an ElementRef
+ * @param view The view to which the node belongs
```
fixed

mhevery commented on packages/core/src/linker/element_ref.ts:
```diff
@@ -6,9 +6,38 @@
+ * @param view The view to which the node belongs
+ * @returns The ElementRef instance to use
+ */
+export function createElementRef(tNode: TNode, view: LView): ElementRef {
+  return new ElementRef(getNativeByTNode(tNode, view) as RElement);
```
fixed

mhevery commented on packages/core/src/linker/template_ref.ts:
```diff
@@ -59,11 +69,63 @@ export abstract class TemplateRef<C> {
+
+const R3TemplateRef = class TemplateRef<T> extends VE_TemplateRef<T> {
+  constructor(
+      private _declarationView: LView, private _declarationTContainer: TContainerNode,
```
fixed

mhevery commented on packages/core/src/linker/template_ref.ts:
```diff
@@ -59,11 +69,63 @@ export abstract class TemplateRef<C> {
+/**
+ * Creates a TemplateRef and stores it on the injector.
+ *
+ * @param TemplateRefToken The TemplateRef type
+ * @param ElementRefToken The ElementRef type
```
fixed

mhevery commented on packages/core/src/linker/template_ref.ts:
```diff
@@ -59,11 +69,63 @@ export abstract class TemplateRef<C> {
+ * @param TemplateRefToken The TemplateRef type
+ * @param ElementRefToken The ElementRef type
+ * @param hostTNode The node on which a TemplateRef is requested
+ * @param hostView The view to which the node belongs
```
fixed

mhevery commented on packages/core/src/linker/template_ref.ts:
```diff
@@ -59,11 +69,63 @@ export abstract class TemplateRef<C> {
+ * @param hostView The view to which the node belongs
+ * @returns The TemplateRef instance or null if we can't create a TemplateRef on a given node type
+ */
+export function createTemplateRef<T>(hostTNode: TNode, hostView: LView): TemplateRef<T>|null {
```
fixed

mhevery commented on packages/core/src/linker/view_container_ref.ts:
```diff
@@ -148,11 +168,266 @@ export abstract class ViewContainerRef {
+  constructor(
+      private _lContainer: LContainer,
+      private _hostTNode: TElementNode|TContainerNode|TElementContainerNode,
+      private _hostView: LView) {
```
fixed

mhevery commented on packages/core/src/linker/view_container_ref.ts:
```diff
@@ -148,11 +168,266 @@ export abstract class ViewContainerRef {
+    const lView = (viewRef as R3ViewRef<any>)._lView!;
+    const tView = lView[TVIEW];
+
+    if (viewRef.destroyed) {
+      throw new Error('Cannot insert a destroyed View in a ViewContainer!');
+    }
```
fixed

mhevery commented on packages/core/src/linker/view_container_ref.ts:
```diff
@@ -148,11 +168,266 @@ export abstract class ViewContainerRef {
+  }
+
+  move(viewRef: ViewRef, newIndex: number): ViewRef {
+    if (viewRef.destroyed) {
+      throw new Error('Cannot move a destroyed View in a ViewContainer!');
+    }
```
fixed

AndrewKushnir reviewed (approved):
LGTM, thanks @mhevery 👍 

(adding "cleanup" label since CI is failing atm)

mhevery commented:
[presubmit](http://test/OCL:341716694:BASE:341750751:1605067656714:7992b84a)

mhevery commented:
[presubmit](http://test/OCL:341716694:BASE:341755537:1605070777162:48151650)

AndrewKushnir commented on packages/core/src/render3/interfaces/container.ts:
```diff
@@ -127,8 +125,11 @@ export interface LContainer extends Array<any> {
    *
    * This is lazily initialized by `ViewContainerRef` when the first view is inserted.
+   *
+   * NOTE: This is stored as `any[]` because render3 should really not be aware of `ViewRef` and
+   * doing so creates circular dependency.
```
Thanks for adding more docs! I'm a bit confused on what the best type (other than `any`) we can use here? It looks like the `VIEW_REFS` field is designed to keep the list of `ViewRef`s - is it something that we should consider changing?

petebacondarwin replied:
I would suggest `unknown` here and then to cast it to `ViewRef` in `getViewRefs()` and `getOrCreateViewRefs()`.

Did you also consider a `WeakMap` to completely remove the connection between `LContainer` and `ViewRef`? Or is that not performant?

AndrewKushnir replied:
Thanks Pete!
+1 to use `unknown` here and I'm still a bit puzzled whether Misko's comment suggests that this field should not be there at all?

petebacondarwin replied:
The `LContainer` is still storing the `ViewRefs` but it no longer "knows" what it is storing.

petebacondarwin replied:
My understanding is that there needs to be a link between these two entities but that we don't really want either to know about the other, which is why I suggested a WeakMap.

mhevery replied:
`WeakMap` would create memory pressure and it would be much slower than current design. The issue is that the ViewEngine wants `ViewRef`s to be persistent, so we need to store them someplace. Not sure how to fix this.

AndrewKushnir commented on packages/core/test/bundling/injection/bundle.golden_symbols.json:
```diff
@@ -5,6 +5,9 @@
   {
```
I believe having these additional symbols in golden files indicate that they are not tree-shakeable now for some reasons. @mhevery what do you think?

mhevery replied:
Hmm. Let me investigate this. Yes this seems wrong.

petebacondarwin reviewed (approved):
Nice tidy up @mhevery !

petebacondarwin reviewed (approved):
Reviewed-for: public-api

jessicajaniuk commented on packages/core/src/linker/template_ref.ts:
```diff
@@ -59,11 +69,61 @@ export abstract class TemplateRef<C> {
+
+    const declarationViewLQueries = this._declarationLView[QUERIES];
+    if (declarationViewLQueries !== null) {
+      embeddedLView[QUERIES] = declarationViewLQueries.createEmbeddedView(embeddedTView);
```
This could be inlined. I don't know what the default value of `embeddedLView[QUERIES]` is. So if it's null, you could just replace after `??` with `null` here.
```
embeddedLView[QUERIES] = declarationViewLQueries?.createEmbeddedView(embeddedTView) ?? embeddedLView[QUERIES];
```

jessicajaniuk commented on packages/core/src/linker/template_ref.ts:
```diff
@@ -59,11 +69,61 @@ export abstract class TemplateRef<C> {
+    ngDevMode && assertDefined(hostTNode.tViews, 'TView must be allocated');
+    return new R3TemplateRef(
+        hostLView, hostTNode as TContainerNode, createElementRef(hostTNode, hostLView));
+  } else {
```
this else is unnecessary .
```suggestion
  }
  return null;
```

jessicajaniuk commented on packages/core/src/linker/view_container_ref.ts:
```diff
@@ -148,11 +168,266 @@ export abstract class ViewContainerRef {
+
+  get(index: number): ViewRef|null {
+    const viewRefs = getViewRefs(this._lContainer);
+    return viewRefs !== null && viewRefs[index] || null;
```
```suggestion
    return viewRefs?.[index] ?? null;
```

jessicajaniuk replied:
It might be better to use `|| null` though depending on what `viewRefs[index]` is.

AndrewKushnir replied:
Similar to https://github.com/angular/angular/pull/39621#discussion_r522431920, there is a size difference from JS perspective (see [TS playground](https://www.typescriptlang.org/play?#code/PTAEFEA8EMFsAcA2BTUBGAXAKCwYwPYB2AzgC6gCWhAVqALygDeWoroAbtIgK7IagByAQBoWbAGbRcpfACcAnv3HdC0ikQAUASiahZyUt1mFBxfLFRlZVAOYCA3KAC+WF1hCgA8vFLqTaUA0AfiC6LWwqagA6Th5UEIZIqMlpOXltexwPb18iUAAmQLpi8Kwk2N56YtAVABNkcSpkWtAgyhoYrkrEjpSZBW1QfnKu5EyssCg4JFR87AIScnYKZAB3ACUG4n5oQnkAbQBdelAjzIWydvrIE4AGceyfP3RAoKjQXZaQ0uW1zfFiG99lRrscQqBCNxEIgHmAcs9ChoAITVSHQn4rDZbUAohhoxCgABkhI4mP+xGBhFBoAAPjSIVCYVggA)):

```
// Option 1 (?. and ??):
(_b = viewRefs === null || viewRefs === void 0 ? void 0 : viewRefs[index]) !== null && _b !== void 0 ? _b : null;

// Option 2 (!== null):
viewRefs !== null && viewRefs[index] || null;
```

In this case the delta is 2x, which is more sensitive from payload size perspective.

jessicajaniuk commented on packages/core/src/linker/view_container_ref.ts:
```diff
@@ -148,11 +168,266 @@ export abstract class ViewContainerRef {
+
+  indexOf(viewRef: ViewRef): number {
+    const viewRefsArr = getViewRefs(this._lContainer);
+    return viewRefsArr !== null ? viewRefsArr.indexOf(viewRef) : -1;
```
```suggestion
    return viewRefsArr?.indexOf(viewRef) ?? -1;
```

JoostK replied:
One downside of the terse syntax here is that it introduce three more comparisons behind the scenes (in the generated JS). One for `viewRefsArray` not being`undefined`, and then the result of `indexOf` against `null` and `undefined` even though that is known to be number.

jessicajaniuk commented on packages/core/src/render3/collect_native_nodes.ts:
```diff
@@ -0,0 +1,85 @@
+        const lViewInAContainer = lNode[i];
+        const lViewFirstChildTNode = lViewInAContainer[TVIEW].firstChild;
+        if (lViewFirstChildTNode !== null) {
+          collectNativeNodes(
```
This recursion within a while loop seems like something that is probably necessary but raises performance concerns for me. Just noting for potential future further cleanup. It's probably unavoidable, but figured it's worth noting.

jessicajaniuk commented on packages/core/src/render3/collect_native_nodes.ts:
```diff
@@ -0,0 +1,85 @@
+
+    const tNodeType = tNode.type;
+    if (tNodeType & TNodeType.ElementContainer) {
+      collectNativeNodes(tView, lView, tNode.child, result);
```
and more recursion here

jessicajaniuk commented on packages/core/src/render3/collect_native_nodes.ts:
```diff
@@ -0,0 +1,85 @@
+            assertDefined(
+                parentView,
+                'Component views should always have a parent view (component\'s host view)');
+        collectNativeNodes(parentView[TVIEW], parentView, nodesInSlot, result, true);
```
and here too.

jessicajaniuk reviewed (approved):
This looks good. I've left a number of comments and suggestions, but they are entirely minor.

mhevery commented on packages/core/src/di/inject_switch.ts:
```diff
@@ -0,0 +1,76 @@
+    token: Type<T>|InjectionToken<T>, notFoundValue: T|undefined, flags: InjectFlags): T|null {
+  const injectableDef: ɵɵInjectableDef<T>|null = getInjectableDef(token);
+  if (injectableDef && injectableDef.providedIn == 'root') {
+    return injectableDef.value === undefined ? injectableDef.value = injectableDef.factory() :
```
I really like your suggestion and I think we should do it. Turns out that our code formatter changes `??=` to `?? =` which creates invalid code. So until we fix that, I can't apply your good suggestion. 😢

AndrewKushnir replied:
Just a quick note: I was curious how this code would look like int JS and here is a [TS playground link](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAlmAVjAvDA3gKBjmA3AQwBsBXAUwC4YByagGm1wDMDgoQAnATyqZLDZxwACgCUGGBzJQSHMDQggAtmRjQOCAObUA3DAC+mQ5kwB6UzADyAByhD5ARhjCA-C5SiKmBIgB0hUlU3NB9fFjZOLjEdE3MrW3sYACZnFDTPbyR-YnJUNJh+ABMyJgQyQpgXeCyA3JCs8PZuMRgqUNqyGKA). Here is a snippet:

```
// Option 1 (??=):
(_a = inj.value) !== null && _a !== void 0 ? _a : (inj.value = inj.factory());

// Option 2 (===):
inj.value === undefined ? inj.value = inj.factory() : inj.value;
```

The `??=` option is slightly larger in size, but from readability perspective I really like Jessica's proposal.

Just wanted to mention that to see if the code size increase is critical (assuming that we'll start using similar approach in other places too).

mhevery commented:
[presubmit](http://test/OCL:341716694:BASE:342142374:1605223236661:b2399a7)

AndrewKushnir reviewed (approved):
LGTM, thanks for all the updates @mhevery 👍

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

IgorMinar reviewed (approved):
This looks awesome! thanks for the cleanup.

Reviewed-for: public-api,  circular-dependencies, size-tracking

AndrewKushnir reviewed (approved):
Reviewed-for: size-tracking

jelbourn reviewed (approved):
Overall LGTM, left a few optional comments on stuff that was moved

Reviewed-for: public-api

mhevery commented on packages/core/src/di/injector_marker.ts:
```diff
@@ -0,0 +1,23 @@
+ *
+ * Currently only `Injector` is special.
+ *
+ * NOTE: the numbers here must be negative.
```
fixed

mhevery commented on packages/core/src/linker/template_ref.ts:
```diff
@@ -59,11 +69,60 @@ export abstract class TemplateRef<C> {
-const SWITCH_TEMPLATE_REF_FACTORY__PRE_R3__ = noop;
-const SWITCH_TEMPLATE_REF_FACTORY: typeof render3InjectTemplateRef =
-    SWITCH_TEMPLATE_REF_FACTORY__PRE_R3__;
+const VE_TemplateRef = TemplateRef;
```
fixed

mhevery commented on packages/core/src/change_detection/change_detector_ref.ts:
```diff
@@ -6,7 +6,18 @@
+import {ViewRef as R3_ViewRef} from '../render3/view_ref';
+import {noop} from '../util/noop';
+
+export const SWITCH_CHANGE_DETECTOR_REF_FACTORY__POST_R3__ = injectChangeDetectorRef;
+const SWITCH_CHANGE_DETECTOR_REF_FACTORY__PRE_R3__ = noop;
+const SWITCH_CHANGE_DETECTOR_REF_FACTORY: typeof injectChangeDetectorRef =
+    SWITCH_CHANGE_DETECTOR_REF_FACTORY__PRE_R3__;
```
This is a pattern we use all over the code base.  (and will be deleted with ViewEngine) Since you marked it as optional I will leave it for a more holistic cleanup for later.

mhevery commented:
[presubmit](http://test/OCL:341716694:BASE:342356557:1605312143801:86b6be85)
